### DISCUSSION TITLE
Matrix questions

### DISCUSSION THREAD

**Discussion by StefanMunnes** [surveydown-dev/surveydown#105] [Feature Requests]
### Description

Hello and thank you for your great work.
This project looks very promising, and we plan to try and implement it in our workflow. But as far as I see, there is the option for a matrix survey missing for now, right? This is a question type for us in the social science used quite often.
Not sure how to implement it in the backend, but could be question type "matrix" and with a vector input for the argument labels = c("Question 1" = subquest_1, "Question 2" = subquest_2) and so on. The options argument should be than the scale (e.g. 1-10). It could also be beneficial if one could specify in an additional option the range of the scale and just provide start and end point labels (e.g. range = c(1,5), options = c("Very clear", "Very unclear"). And one last point, having an additional and somewhat separated "Don't know" option in the end would also be great.

Hope you think this is useful and easy to implement. And again, a big Thank You!

**Comment by jhelvy** [comment]
Yes this is definitely on our development radar as I know it is a common question type people need. I saw that a somewhat similar (though older) package that also leverages shiny ([shinysurveys](https://cran.r-project.org/web/packages/shinysurveys/vignettes/surveying-shinysurveys.html)) has a matrix question type, so I know this is do-able in shiny. We just need to sit down and figure out how to implement it. Now that we've addressed a lot of more pressing issues (e.g. security & architectural changes in v0.3.0), we'll put this back on our main todo list. 

@pingfan-hu if you get some time could you look into it? I think searching through how it was implemented in [shinysurveys](https://github.com/jdtrat/shinysurveys) could be helpful.

**Comment by pingfan-hu** [comment]
Hi @StefanMunnes ! The matrix question type is done! You can see the documentation [here](https://surveydown.org/question-types#matrix), and try the demo survey [here](https://surveydown.shinyapps.io/types/).
The demo survey is just a showcase of all question types, including matrix and all other types.
<img width="561" alt="image" src="https://github.com/user-attachments/assets/b86655c5-6b1c-4791-a12e-414308f216e9">

**Reply by StefanMunnes** [reply]
@pingfan-hu Wow, that was fast. Thank you. I try to test this new question type as soon as possible.
Just for now, in the demo survey, the matrix question is missing in the overview list of the question types.

**Reply by pingfan-hu** [reply]
Oops, good catch. I'll fix it today.
Yeah please update to our latest version and have a try.

**Reply by pingfan-hu** [reply]
Done. The `matrix` type is on the list now.

**Comment by StefanMunnes** [comment]
Two questions:

1. is there a logical reason why the label of the subquestion and the "corresponding ID" in the row option are the other way around than in the answer options? In the answer options the label is written on the left side of the equal sign, in the row option it has to be written on the right side. Intuitively, I had it parallel to the answer options in the first attempt, and that's how the IDs were displayed.

2. in the Roxygen help it is written that you have to specify a list, like in option, but in the example you specify a named vector for option and a named list for row. In my test, I also intuitively specified a vector with row = c("ID" = "label") without an error occurring. Can you explain in more detail why what is necessary?

**Comment by jhelvy** [comment]
I agree, it is perhaps a bit confusing to have order swapped around. @pingfan-hu let's discuss. Here's an example to clarify the comment. Our current design is this:

```r
sd_question(
  type   = "matrix",
  id     = "car_preference",
  label  = "Please indicate your agreement with the following statements.",
  row    = c(
    "buy_gasoline" = "I'd like to buy a gasoline car",
    "buy_ev"       = "I'd like to buy an EV"
  ),
  option = c(
    "Disagree" = "disagree",
    "Neutral"  = "neutral",
    "Agree"    = "agree"
  )
)
```

In the `option` argument, the label is on the left, and value is on the right. The `row` argument has the opposite order. I think I know why you did this though: 1) it is easier to work with resulting object in the code if `row` is set up this way, and 2) the label is almost always going to be longer, so it's a bit awkward to put it in the other order. 

This is what the same question would look like if the order was reversed:

```r
sd_question(
  type   = "matrix",
  id     = "car_preference",
  label  = "Please indicate your agreement with the following statements.",
  row    = c(
    "I'd like to buy a gasoline car" = "buy_gasoline",
    "I'd like to buy an EV" = "buy_ev"
  ),
  option = c(
    "Disagree" = "disagree",
    "Neutral"  = "neutral",
    "Agree"    = "agree"
  )
)
```

I have to agree here with @StefanMunnes, this does feel more consistent to write this way. It should be simple enough to swap around how this works in the code if you agree.

As for the Roxygen question, that is probably just a typo. @pingfan-hu can you double check? I think it should say "named vector" instead of list.

**Comment by pingfan-hu** [comment]
Hi folks, yeah I'll fix the issues. I didn't intentionally define the row option is the other way around. I simply didn't notice the mistake. @jhelvy Professor you are right: it's better to write this way. I'll update the source code.
And yeah, I'll update the Roxygen document.

**Comment by pingfan-hu** [comment]
Hi @StefanMunnes ! It's now fixed. An example now looks like this:
```r
sd_question(
  type   = "matrix",
  id     = "car_preference",
  label  = "Please indicate your agreement with the following statements.",
  row    = c(
    "I'd like to buy a gasoline car." = "buy_gasoline",
    "I'd like to buy an EV." = "buy_ev"
  ),
  option = c(
    "Disagree" = "disagree",
    "Neutral"  = "neutral",
    "Agree"    = "agree"
  )
)
```

As you can see, the `row` parameter now has labels on the left, and values on the right.
The Roxygen document is updated as well.

Also, I noticed a glitch that the base matrix question id appears in the data sheet. Now it's fixed.
For example, in the previous version, for the above example we have base question id being `car_preference`, and the combined sub-question ids are `car_preference_buy_gasoline` and `car_preference_buy_ev`, and all 3 ids appear in the data sheet. But the problem is, there is no such `car_preference` question.
Now I've fixed the issue, so that the base id never appears and only the 2 sub-question ids are shown.

Please update to the latest version using the code below and have a test:
```r
remotes::install_github("surveydown-dev/surveydown", force = TRUE)
```

**Reply by StefanMunnes** [reply]
Looks great, thank you!

**Reply by StefanMunnes** [reply]
It seems the change in the appearance of the base question ID leads to a new errror in the sd_show_if() function. I receive the error: 
`Invalid show_if targets: date_new. These must be question IDs defined in the survey.qmd file.```

For me, it looks like, it's not possible anymore to address the whole question by the ID.

**Reply by jhelvy** [reply]
Hmm, this is an interesting issue. I'm not actually sure what the appropriate interface should be for specifying a show_if condition for a matrix type question. Since a matrix question is really a set of multiple questions, using the root question `id` as part of a show_if condition seems like it wouldn't make sense, as we can't know which specific question was answered. Using the same `car_preference` example above, I would expect you should need to specify the sub-question in the show_if, like this:

```r
sd_show_if(
  input$car_preference_buy_gasoline == "disagree" ~ "target_question_id"
)
```

If I tried to just use `car_preference` in the condition, there is no choice option for that `id`, so I don't know what I would specify. I think we might just need to clarify in the documentation that for matrix type questions we generate an `id` for each sub-question in the `row` argument, which takes the form `question_id_row_id`.

@StefanMunnes how were you thinking of using the root question `id` for a matrix type question? Does this make sense?

**Reply by StefanMunnes** [reply]
That's a good point, but I was struggling with the other way around:
```r
sd_show_if(
    input$interessted_buying_car == "yes" ~ "car_preference"
)
```
I understand the logic why it could make sense to address the single sub-questions, that is definitively useful. But in most cases it should also be possible to reduce the amount of conditions if you want to navigate to the whole questions by the base ID ("car_preference") instead of every single sub-question (e.g. "car_preference_ev" ...).

The other way around, as you presented in your example, is totally fine. In some cases, I would filter in show_if(), when all the subquestions are empty (e.g. `length(input$car_preference_buy_gasoline) == 0 && length(input$car_preference_buy_ev) == 0`), not sure if you have a shortcut for this case?

**Reply by jhelvy** [reply]
Ah I see. Yes, we should be able to use the root `id` for a matrix question if it is the _target_ of some condition, meaning the entire matrix question will display if some condition is met. @pingfan-hu can you look into both of these situations? That is, 1) allow the root matrix question id to be used as show_if _targets_, and 2) allow matrix sub questions to be used as show_if _conditions_. We should update the roxygen as well, maybe using the "details" tag in sd_question to describe this (could also update the question types docs page).

As for the point about testing if all subquestions are empty, I opened a new feature request [here](https://github.com/orgs/surveydown-dev/discussions/122)

**Reply by pingfan-hu** [reply]
OK, @StefanMunnes , now it's solved. The reason why it used to work and then didn't is because that I removed the root question id from the data sheet. Now it's added back.
I removed it because this root question id will never be answered and will prohibit you from moving on to the next page, if you set all questions being required. My solution is that I eliminated the matrix root question ids from the required logic.

Long story short: you can again apply the root question id, and if you define all questions as required, there won't be any error.

An implementation of `sd_show_if()` is like this (same as how you desired):

```r
sd_show_if(
    input$interested_buying_car == "yes" ~ "car_preference"
)
```

Please update our package and try it out:

```r
remotes::install_github("surveydown-dev/surveydown", force = TRUE)
```

**Comment by jhelvy** [comment]
Closing as this is now supported
